### PR TITLE
Stats: Fixed variance of sum of Poisson Random Variables  when parameter is a float

### LINKED ISSUE #20669: Cannot find variance of sum of Poisson distribution Random Variables when the parameter is a float
I am using Sympy 1.7.1.

I am trying to find the variance of the sum of two poisson random variables. This is the code I tried:

```python
from sympy.stats import variance, Poisson
import sympy

sympy.__version__

X1 = Poisson("X1", 5.2)
X2 = Poisson("X2", 5.2)

# this works
variance(X1), variance(X2)

# this fails
variance(X1 + X2)
```

The last line throws the following exception:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-21-fc81032a19ce> in <module>
----> 1 variance(X1 + X2)

~/sage/local/lib/python3.8/site-packages/sympy/stats/rv_interface.py in variance(X, condition, **kwargs)
     71         return Variance(X, condition)
     72 
---> 73     return cmoment(X, 2, condition, **kwargs)
     74 
     75 

~/sage/local/lib/python3.8/site-packages/sympy/stats/rv_interface.py in cmoment(X, n, condition, evaluate, **kwargs)
    227     from sympy.stats.symbolic_probability import CentralMoment
    228     if evaluate:
--> 229         return CentralMoment(X, n, condition).doit()
    230     return CentralMoment(X, n, condition).rewrite(Integral)
    231 

~/sage/local/lib/python3.8/site-packages/sympy/stats/symbolic_probability.py in doit(self, **hints)
    667         if not is_random(self.args[0]):
    668             return self.args[0]
--> 669         return self.rewrite(Expectation).doit(**hints)
    670 
    671     def _eval_rewrite_as_Expectation(self, X, n, condition=None, **kwargs):

~/sage/local/lib/python3.8/site-packages/sympy/stats/symbolic_probability.py in doit(self, **hints)
    273             return self.func(expr)
    274         # Otherwise case is simple, pass work off to the ProbabilitySpace
--> 275         result = pspace(expr).compute_expectation(expr, evaluate=for_rewrite)
    276         if hasattr(result, 'doit') and for_rewrite:
    277             return result.doit(**hints)

~/sage/local/lib/python3.8/site-packages/sympy/stats/rv.py in compute_expectation(self, expr, rvs, evaluate, **kwargs)
    417             expr = space.compute_expectation(expr, rvs & space.values, evaluate=False, **kwargs)
    418         if evaluate and hasattr(expr, 'doit'):
--> 419             return expr.doit(**kwargs)
    420         return expr
    421 

~/sage/local/lib/python3.8/site-packages/sympy/concrete/summations.py in doit(self, **hints)
    191     def doit(self, **hints):
    192         if hints.get('deep', True):
--> 193             f = self.function.doit(**hints)
    194         else:
    195             f = self.function

~/sage/local/lib/python3.8/site-packages/sympy/core/operations.py in doit(self, **hints)
    430     def doit(self, **hints):
    431         if hints.get('deep', True):
--> 432             terms = [term.doit(**hints) for term in self.args]
    433         else:
    434             terms = self.args

~/sage/local/lib/python3.8/site-packages/sympy/core/operations.py in <listcomp>(.0)
    430     def doit(self, **hints):
    431         if hints.get('deep', True):
--> 432             terms = [term.doit(**hints) for term in self.args]
    433         else:
    434             terms = self.args

~/sage/local/lib/python3.8/site-packages/sympy/concrete/summations.py in doit(self, **hints)
    204         if reps:
    205             undo = {v: k for k, v in reps.items()}
--> 206             did = self.xreplace(reps).doit(**hints)
    207             if type(did) is tuple:  # when separate=True
    208                 did = tuple([i.xreplace(undo) for i in did])

~/sage/local/lib/python3.8/site-packages/sympy/concrete/summations.py in doit(self, **hints)
    230                 f = -f
    231 
--> 232             newf = eval_sum(f, (i, a, b))
    233             if newf is None:
    234                 if f == self.function:

~/sage/local/lib/python3.8/site-packages/sympy/concrete/summations.py in eval_sum(f, limits)
    996     # this can save time when b-a is big.
    997     # We should try to transform to partial fractions
--> 998     value = eval_sum_symbolic(f.expand(), (i, a, b))
    999     if value is not None:
   1000         return value

~/sage/local/lib/python3.8/site-packages/sympy/concrete/summations.py in eval_sum_symbolic(f, limits)
   1118             # Try term by term
   1119             lsum = eval_sum_symbolic(L, (i, a, b))
-> 1120             rsum = eval_sum_symbolic(R, (i, a, b))
   1121 
   1122             if None not in (lsum, rsum):

~/sage/local/lib/python3.8/site-packages/sympy/concrete/summations.py in eval_sum_symbolic(f, limits)
   1118             # Try term by term
   1119             lsum = eval_sum_symbolic(L, (i, a, b))
-> 1120             rsum = eval_sum_symbolic(R, (i, a, b))
   1121 
   1122             if None not in (lsum, rsum):

~/sage/local/lib/python3.8/site-packages/sympy/concrete/summations.py in eval_sum_symbolic(f, limits)
   1117         else:
   1118             # Try term by term
-> 1119             lsum = eval_sum_symbolic(L, (i, a, b))
   1120             rsum = eval_sum_symbolic(R, (i, a, b))
   1121 

~/sage/local/lib/python3.8/site-packages/sympy/concrete/summations.py in eval_sum_symbolic(f, limits)
   1077         without_i, with_i = f.as_independent(i)
   1078         if without_i != 1:
-> 1079             s = eval_sum_symbolic(with_i, (i, a, b))
   1080             if s:
   1081                 r = without_i*s

~/sage/local/lib/python3.8/site-packages/sympy/concrete/summations.py in eval_sum_symbolic(f, limits)
   1195             return r
   1196 
-> 1197     h = eval_sum_hyper(f_orig, (i, a, b))
   1198     if h is not None:
   1199         return h

~/sage/local/lib/python3.8/site-packages/sympy/concrete/summations.py in eval_sum_hyper(f, i_a_b)
   1298 
   1299     # Now b == oo, a != -oo
-> 1300     res = _eval_sum_hyper(f, i, a)
   1301     if res is not None:
   1302         r, c = res

~/sage/local/lib/python3.8/site-packages/sympy/concrete/summations.py in _eval_sum_hyper(f, i, a)
   1217         if simplify(f.subs(i, Dummy('i', integer=True, positive=True))) == 0:
   1218             return S.Zero, True
-> 1219         return _eval_sum_hyper(f.subs(i, i + 1), i, 0)
   1220 
   1221     hs = hypersimp(f, i)

~/sage/local/lib/python3.8/site-packages/sympy/concrete/summations.py in _eval_sum_hyper(f, i, a)
   1255     h = hyper(ap, bq, x)
   1256     f = combsimp(f)
-> 1257     return f.subs(i, 0)*hyperexpand(h), h.convergence_statement
   1258 
   1259 

~/sage/local/lib/python3.8/site-packages/sympy/simplify/hyperexpand.py in hyperexpand(f, allow_hyper, rewrite, place)
   2489         if not r.has(nan, zoo, oo, -oo):
   2490             return r
-> 2491     return f.replace(hyper, do_replace).replace(meijerg, do_meijer)

~/sage/local/lib/python3.8/site-packages/sympy/core/basic.py in replace(self, query, value, map, simultaneous, exact)
   1491             return expr
   1492 
-> 1493         rv = walk(self, rec_replace)
   1494         return (rv, mapping) if map else rv
   1495 

~/sage/local/lib/python3.8/site-packages/sympy/core/basic.py in walk(rv, F)
   1475                                 if rv == e and e != newargs[i]:
   1476                                     return rv
-> 1477                 rv = F(rv)
   1478             return rv
   1479 

~/sage/local/lib/python3.8/site-packages/sympy/core/basic.py in rec_replace(expr)
   1484             result = _query(expr)
   1485             if result or result == {}:
-> 1486                 v = _value(expr, result)
   1487                 if v is not None and v != expr:
   1488                     if map:

~/sage/local/lib/python3.8/site-packages/sympy/core/basic.py in <lambda>(expr, result)
   1413                 _value = lambda expr, result: value(*expr.args)
   1414             elif callable(value):
-> 1415                 _value = lambda expr, result: value(*expr.args)
   1416             else:
   1417                 raise TypeError(

~/sage/local/lib/python3.8/site-packages/sympy/simplify/hyperexpand.py in do_replace(ap, bq, z)
   2478 
   2479     def do_replace(ap, bq, z):
-> 2480         r = _hyperexpand(Hyper_Function(ap, bq), z, rewrite=rewrite)
   2481         if r is None:
   2482             return hyper(ap, bq, z)

~/sage/local/lib/python3.8/site-packages/sympy/simplify/hyperexpand.py in _hyperexpand(func, z, ops0, z0, premult, prem, rewrite)
   2017     # Try to recognise a shifted sum.
   2018     p = S.Zero
-> 2019     res = try_shifted_sum(func, z0)
   2020     if res is not None:
   2021         func, nops, p = res

~/sage/local/lib/python3.8/site-packages/sympy/simplify/hyperexpand.py in try_shifted_sum(func, z)
   1671 
   1672     ops = []
-> 1673     for n in range(r - 1):
   1674         ops.append(ShiftA(n + 1))
   1675     ops.reverse()

TypeError: 'Float' object cannot be interpreted as an integer
```

The code snippet works as expected if the parameter to `Poisson()` is an integer e.g. 5.

### PR DESCRIPTION
**Pull request by prototypevito**
#### References to other Issues or PRs

Fixes #20669

#### Brief description of what is fixed or changed
-> This commit does not include the type change of  r to int but solves the problem that followed: the code returning an unevaluated limit
-> The problem lay in the summations.py file present in the concrete module, where the existing code could only detect if the number was a float. It could not check for float numbers present in a fraction like (7.5)/x+1.  The same has been fixed.
#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by prototypevito** [comment]
This commit also fixes the issue: Finding the expected value of the product of two random variables with floating type parameters. 
How I arrived at the solution? 
Upon comparing the two cases (using an integer as the parameter for Poisson and then using a floating type), the workflow was identical for both **until**  `numer, denom = fraction(factor(hs))` in **summations.py** at which the float type case could not be factorized. This should have been taken care of by the line `if isinstance(hs, Float)` but the instance function returns False for cases where only a part of the input is a float. 
For example:
`if isinstance(5.0, Float)` gives `True` 
but
`if isinstance(5.0/(x+1), Float)` gives `False`, x being a variable

Hence the change proposed works. @czgdp1807 let me know if this is the right way of solving this issue.

**Review by czgdp1807 — CHANGES_REQUESTED** [review]
Please add a regression test for the fix. Then I would be able to review it more thoroughly. Thanks for sending the PR. :-)

**Comment by prototypevito** [comment]
> Please add a regression test for the fix. Then I would be able to review it more thoroughly. Thanks for sending the PR. :-)

Hey, I have added a test. Let me know if the test is appropriate.

File: sympy/stats/tests/test_rv.py [hunk 1]
```diff
@@ -439,3 +439,12 @@ def test_issue_20286():
...
     assert eq.dummy_eq(H(B))
+
+def test_issue_20669():
+    from sympy import N
+    X1 = Poisson("X1", 5.2)
+    X2 = Poisson("X2", 5.2)
+    X3 = Poisson("X3", 5)
+    assert N(variance(X1+X2),7)==N(variance(X1)+variance(X2),7)
+    assert N(E(X1*X2),7)==N(E(X1)*E(X2),7)
+    assert N(E(X1*X3),7)==N(E(X1)*E(X3),7)
```
Comment (czgdp1807) [inline sympy/stats/tests/test_rv.py:450]:
```suggestion
    assert N(variance(X1 + X2), 7) == N(variance(X1) + variance(X2), 7)
    assert N(E(X1*X2), 7) == N(E(X1) * E(X2), 7)
    assert N(E(X1*X3), 7) == N(E(X1) * E(X3), 7)
```

**Review by czgdp1807 — APPROVED** [review]
Other than this looks good to me. @Upabjojr thoughts?

**Comment by prototypevito** [comment]
@oscarbenjamin I have added my Name and email ID to the authors file, but it still gives a test error. Is it because none of my pull requests have been merged yet? Because when I run mailmap_check.py on my local device, it does not give me an error.

**Comment by prototypevito** [comment]
> Other than this looks good to me. @Upabjojr thoughts?

@Upabjojr Kindly let me know if there is anything else to be changed. Thanks!

**Pull request closed without merging**